src/OS/QMidi_ALSA.cpp

```diff
@@ -118,4 +126,233 @@ void QMidiOut::sendMsg(qint32 msg)
 
 	snd_seq_event_output(fMidiPtrs->midiOutPtr, &ev);
 	snd_seq_drain_output(fMidiPtrs->midiOutPtr);
+
+	snd_midi_event_free(mev);
+}
+
+void QMidiOut::sendSysEx(const QByteArray &data)
+{
+    if (!fConnected)
+    {
+        return;
+    }
+
+    snd_seq_event_t ev;
+    snd_midi_event_t* mev;
+
+    snd_seq_ev_set_source(&ev, 0);
+    snd_seq_ev_set_subs(&ev);
+    snd_seq_ev_set_direct(&ev);
+
+    snd_midi_event_new(data.size(), &mev);
+    snd_midi_event_resize_buffer(mev, data.size());
+    snd_midi_event_encode(mev, (unsigned char*) data.data(), data.size(), &ev);
+
+    snd_seq_event_output(fMidiPtrs->midiOutPtr, &ev);
+    snd_seq_drain_output(fMidiPtrs->midiOutPtr);
+
+    snd_midi_event_free(mev);
+}
+
+/* ------------------------------[ QMidiIn ]------------------------------ */
+
+#include "QMidiIn.h"
+#include <QByteArray>
+
+#include "OS/QMidi_ALSA.h"
+
+struct NativeMidiInInstances {
+    //!
+    //! \brief midiIn is a reference to the MIDI input device
+    //!
+    snd_seq_t* midiIn;
+
+    //!
+    //! \brief receiveThread is a reference to the MIDI input receive thread.
+    //!
+    MidiInReceiveThread* receiveThread;
+};
+
+QMap<QString, QString> QMidiIn::devices()
+{
+    QMap<QString, QString> ret;
+
+    snd_seq_client_info_t* cinfo;
```

Comment:
This code seems mostly duplicated with QMidiOut::devices(). Could we perhaps combine the two via a (private) utility function?